auth.php

```diff
@@ -323,10 +338,88 @@ protected function is_ip_restriction_enabled() {
      * @return object User object.
      */
     protected function create_user(array $data) {
-        // TODO:
-        // 1. Validate user
-        // 2. Create user.
-        // 3. Throw exception if something went wrong.
+        global $DB, $CFG;
+
+        $user = $data;
+        unset($user['ip']);
+        $user['auth'] = 'userkey';
+        $user['mnethostid'] = $CFG->mnet_localhost_id;
+
+        $requiredfieds = ['username', 'email', 'firstname', 'lastname'];
+        $missingfields = [];
+        foreach ($requiredfieds as $requiredfied) {
+            if (empty($user[$requiredfied])) {
+                $missingfields[] = $requiredfied;
+            }
+        }
+        if (!empty($missingfields)) {
+            throw new invalid_parameter_exception('Unable to create user, missing value(s): ' . implode(',', $missingfields));
+        }
+
+        if ($DB->record_exists('user', array('username' => $user['username'], 'mnethostid' => $CFG->mnet_localhost_id))) {
+            throw new invalid_parameter_exception('Username already exists: '.$user['username']);
+        }
+        if (!validate_email($user['email'])) {
+            throw new invalid_parameter_exception('Email address is invalid: '.$user['email']);
+        } else if (empty($CFG->allowaccountssameemail) &&
+            $DB->record_exists('user', array('email' => $user['email'], 'mnethostid' => $user['mnethostid']))) {
+            throw new invalid_parameter_exception('Email address already exists: '.$user['email']);
+        }
+
+        $userid = user_create_user($user);
+        return $DB->get_record('user', ['id' => $userid]);
+    }
+
+    /**
+     * Update an existing user.
+     *
+     * @param stdClass $user Existing user record.
+     * @param array $data Validated user data from web service.
+     *
+     * @return object User object.
+     */
+    protected function update_user(\stdClass $user, array $data) {
```

Comment:
There's not really much to gain, this function does very little since we are only handling a tiny proportion of the fields on the user. It also imposes a lot of it's own validation which is either not relevant or would be a problem in this very specific use case.